pkg/k8s/identitybackend/identity.go

```diff
@@ -0,0 +1,422 @@
+// we map the labels directly to the ciliumidentity CRD instance, and
+// kubernetes does not allow ':' in the name of the label.
+func stripK8sPrefix(old map[string]string) map[string]string {
+	fixup := make(map[string]string, len(old))
+	for k, v := range old {
+		k = strings.TrimPrefix(k, "k8s:")
+		fixup[k] = v
+	}
+	return fixup
+}
+
+// AllocateID will create an identity CRD, thus creating the identity for this
+// key-> ID mapping.
+// Note: This does not create a reference to this node to indicate that it is
+// using this identity. That must be done with AcquireReference.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) AllocateID(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) error {
+	identity := &v2.CiliumIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   id.String(),
+			Labels: stripK8sPrefix(key.GetAsMap()),
+		},
+		Status: v2.IdentityStatus{
+			Nodes: map[string]metav1.Time{
+				c.NodeName: metav1.Now(),
+			},
+		},
+	}
+
+	_, err := c.Client.CiliumV2().CiliumIdentities().Create(identity)
+	return err
+}
+
+func (c *crdBackend) AllocateIDIfLocked(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	return c.AllocateID(ctx, id, key)
+}
+
+// JSONPatch structure based on the RFC 6902
+type JSONPatch struct {
+	OP    string      `json:"op,omitempty"`
+	Path  string      `json:"path,omitempty"`
+	Value interface{} `json:"value"`
+}
+
+// AcquireReference updates the status field of the CRD corresponding to id
+// with this node. This marks that CRD as used by this node, and will stop it
+// being garbage collected.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) AcquireReference(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	identity := c.get(ctx, key)
+	if identity == nil {
+		return fmt.Errorf("identity does not exist")
+	}
+
+	capabilities := k8sversion.Capabilities()
+	identityOps := c.Client.CiliumV2().CiliumIdentities()
+
+	var err error
+	if capabilities.Patch {
+		var patch []byte
+		patch, err = json.Marshal([]JSONPatch{
+			{
+				OP:    "test",
+				Path:  "/status",
+				Value: nil,
+			},
+			{
+				OP:   "add",
+				Path: "/status",
+				Value: v2.IdentityStatus{
+					Nodes: map[string]metav1.Time{
+						c.NodeName: metav1.Now(),
+					},
+				},
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		_, err = identityOps.Patch(identity.GetName(), k8sTypes.JSONPatchType, patch, "status")
+		if err != nil {
+			patch, err = json.Marshal([]JSONPatch{
+				{
+					OP:    "replace",
+					Path:  "/status/nodes/" + c.NodeName,
+					Value: metav1.Now(),
+				},
+			})
+			if err != nil {
+				return err
+			}
+			_, err = identityOps.Patch(identity.GetName(), k8sTypes.JSONPatchType, patch, "status")
+		}
+
+		if err == nil {
+			return nil
+		}
+		log.WithError(err).Debug("Error patching status. This is not fatal")
```

Comment:
You should invert the if statement above and log the error. I know the patching status is not fatal but we then should return if err != nil and not continue to execute the rest of the function.

Reply:
Oh I got it, you are continuing because you want to fallback to create the endpoint status with UpdateStatus and / or Update. Please add a comment describing that.